src/librustc_resolve/lib.rs

```diff
@@ -4796,12 +4786,25 @@ impl<'a, 'crateloader: 'a> Resolver<'a, 'crateloader> {
                 err.span_suggestion_with_applicability(
                     binding.span,
                     rename_msg,
-                    if snippet.ends_with(';') {
-                        format!("{} as {};", &snippet[..snippet.len() - 1], suggested_name)
-                    } else {
-                        format!("{} as {}", snippet, suggested_name)
+                    match (
+                        snippet.split_whitespace().find(|w| *w == "as"),
+                        snippet.ends_with(";")
+                    ) {
+                        (Some(_), false) => format!("{} as {}",
+                            &snippet[..snippet.find(" as ").unwrap()],
+                            suggested_name,
+                        ),
+                        (Some(_), true) => format!("{} as {};",
+                            &snippet[..snippet.find(" as ").unwrap()],
```

Comment:
Here (and above) this is not going to work if the snippet is `"use\tfoo\tas\tbar;"`.

----

I'm looking at [`ImportDirectiveSubclass`](https://github.com/rust-lang/rust/blob/bef62ccddb911b9cd7677717a69348a62cf61f96/src/librustc_resolve/resolve_imports.rs#L43-L57) and what `Ident` already has a `span` field, so we can use `target.span` for the single `use` case. For `extern crate`, we should modify [`ImportDirectiveSubclass::ExternCrate(Option<Name>)`](https://github.com/rust-lang/rust/blob/bef62ccddb911b9cd7677717a69348a62cf61f96/src/librustc_resolve/resolve_imports.rs#L55) to be `ImportDirectiveSubclass::ExternCrate(Option<Ident>)`. By doing that you won't have to peek into the snippet to find the appropriate `Span` to use in the suggestion, nor peek into the span to be able to tell wether there's an `as` rename happening. Changing `ExternCrate` will have some knock down effects, but they should be minor.